pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,494 @@
+			t.logger.Log(
+				logging.LevelError, fmt.Sprintf("%s failed to open stream to %s: %v", t.ownID, dest, streamErr))
+		} else {
+			t.logger.Log(
+				logging.LevelInfo, fmt.Sprintf("%s failed to open stream to %s: %v", t.ownID, dest, streamErr))
+		}
+
+		delay := time.NewTimer(retryTimeout)
+		select {
+		case <-delay.C:
+			continue
+		case <-ctx.Done():
+			if !delay.Stop() {
+				<-delay.C
+			}
+			return nil, fmt.Errorf("%s opening stream to %s: context closed", t.ownID, dest)
+		}
+	}
+	return nil, fmt.Errorf("%s failed to open stream to %s: %w", t.ownID, dest, streamErr)
+}
+
+func (t *Transport) Send(ctx context.Context, dest types.NodeID, payload *messagepb.Message) error {
+	var nodeErr *nodeInfoError
+	var streamErr *streamInfoError
+	s, err := t.getNodeStream(dest)
+	if errors.As(err, &nodeErr) {
+		return err
+	}
+	if errors.As(err, &streamErr) {
+		t.connWg.Add(1)
+		go t.connectToNode(ctx, dest, t.connWg)
+		return err
+	}
+
+	outBytes, err := proto.Marshal(payload)
+	if err != nil {
+		return fmt.Errorf("failed to marshal message: %w", err)
+	}
+
+	msg := TransportMessage{t.ownID.Pb(), outBytes}
+	buf := new(bytes.Buffer)
+	err = msg.MarshalCBOR(buf)
+	if err != nil {
+		return fmt.Errorf("failed to CBOR marshal message: %w", err)
+	}
+
+	if ctx.Err() != nil {
+		return err
+	}
+
+	w := bufio.NewWriter(s)
+	_, err = w.Write(buf.Bytes())
+	if err == nil {
+		if err = w.Flush(); err != nil {
+			return fmt.Errorf("failed to flush: %w", err)
+		}
+	}
+	// If we cannot write to the stream, or we don't have a connection to that peer then try to
+	// connect to the peer.
+	if err != nil || len(t.host.Network().ConnsToPeer(s.Conn().RemotePeer())) == 0 {
+		t.connWg.Add(1)
+		go t.connectToNode(ctx, dest, t.connWg)
+		return fmt.Errorf("%s failed to write data to stream for %s, reopening", t.ownID, dest)
+	}
+
+	return nil
+}
+
+func (t *Transport) mirHandler(s network.Stream) {
+	t.logger.Log(logging.LevelDebug, "mir handler started", "src", t.ownID, "dst", s.ID())
+	defer t.logger.Log(logging.LevelDebug, "mir handler stopped", "src", t.ownID, "dst", s.ID())
+
+	defer func() {
+		t.logger.Log(logging.LevelDebug, "mir handler is closing stream", "src", t.ownID, "dst", s.ID())
+		err := s.Close()
+		if err != nil {
+			t.logger.Log(logging.LevelError, "closing stream", "src", t.ownID, "dst", s.ID(), "err", err)
+		}
+	}() // nolint
+
+	for {
+		var msg TransportMessage
+		err := msg.UnmarshalCBOR(s)
+		if err != nil {
+			t.logger.Log(logging.LevelError, "failed to read mir request", "src", t.ownID, "dst", s.ID(), "err", err)
+			return
+		}
+
+		var payload messagepb.Message
+
+		if err := proto.Unmarshal(msg.Payload, &payload); err != nil {
+			t.logger.Log(logging.LevelError, "failed to unmarshall mir request", "src", t.ownID, "dst", s.ID(), "err", err)
+			return
+		}
+
+		t.incomingMessages <- events.ListOf(
+			events.MessageReceived(types.ModuleID(payload.DestModule), types.NodeID(msg.Sender), &payload),
+		)
```

Comment:
What if the destination module does not exist?

Reply:
Do you mean what if the destination module doesn't read from the channel?
Or what will Mir do if it gets a message to a module that doesn't exist?

Reply:
The latter

Reply:
@matejpavlovic Do you see any problems with that? I think that the network transport should know about that and assume that Mir properly handles that.

Reply:
Yes, this is not the Net module's responsibility. The node must handle the situation. This is part of input sanitization.
While complex input is a bigger issue, I've just addressed this particular part in #253 . So I think this comment can be marked as resolved.

Reply:
I think one can make an argument that it's the module's fault emitting events which target a wrong module.